src/main/java/org/testng/internal/PackageUtils.java

```diff
@@ -111,9 +111,9 @@ else if ("jar".equals(protocol)) {
               packageName = name.substring(0, idx).replace('/', '.');
             }
 
-            Utils.log("PackageUtils", 4, "Package name is " + packageName);
-            if ((idx != -1) || recursive) {
+            if (recursive || packageName.equals(packageOnly)) {
```

Comment:
'idx' was used to replace '/' in dir names to  '.' (to become packageName).
There is no chance to detect whether this file is in subpackage or note. We should compare packages (parent directory names)